### PR TITLE
feat: add turn-level prompt caching + PostHog metrics + live API tests

### PR DESCRIPTION
**Pull request by sestinj**
## Summary

- **Turn-level caching**: The `systemAndTools` caching strategy now calls `addCacheControlToLastTwoUserMessages()` to cache conversation turns, not just the system+tools prefix. This matches what the VS Code extension path already does.
- **PostHog cache metrics**: `recordStreamTelemetry` now reports `prompt_cache_metrics` events with `cache_read_tokens`, `cache_write_tokens`, `total_prompt_tokens`, `cache_hit_rate`, and `tool_count`.
- **Live API tests**: 13 tests across 2 files validate caching end-to-end against the real Anthropic API (guarded by `ANTHROPIC_API_KEY` env var, skipped in CI).

### Impact of turn-level caching (validated by temporarily reverting the fix)

| Scenario | Without fix | With fix | Improvement |
|---|---|---|---|
| Basic 3-turn conversation | 93% | 99.4% | +6% |
| Tool use follow-up | 81.9% | 99.5% | +18% |
| Parallel tool calls | 83.5% | 99.4% | +16% |
| Long conversation (8 turns) | 77.5% | 94.0% | +17% |
| Large tool result (~200 lines) | 62.3% | 99.4% | **+37%** |

Without the fix, cache hit rates degrade as conversations grow because only the static system+tools prefix is cached. With turn-level caching, the last two user messages are also cached, keeping rates at 94-99%.

## Test plan

- [x] All 25 existing `AnthropicCachingStrategies.test.ts` unit tests pass (+ 2 new)
- [x] 3 live API tests in `anthropic-caching.live.test.ts` pass
- [x] 10 battle test scenarios in `anthropic-caching-scenarios.live.test.ts` pass
- [x] Validated fix impact by temporarily reverting and comparing hit rates

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

**Continue Tasks:** ❌ 7 failed — [View all](https://hub.continue.dev/inbox/pr/continuedev/continue/10801?utm_source=github_pr&utm_medium=pr_body&utm_campaign=continue_tasks)

---
## Summary by cubic
Adds turn-level prompt caching to the Anthropic systemAndTools strategy, records prompt cache metrics to PostHog, and adds live Anthropic API tests to keep cache hit rates high in multi-turn chats. Also makes the caching transform non‑mutating, fixes minor test/telemetry issues, and ensures live API tests are excluded from CI.

- **New Features**
  - Turn-level caching: caches the last two user messages (plus system + tools).
  - PostHog telemetry: records prompt_cache_metrics (cache_read/write_tokens, total_prompt_tokens, cache_hit_rate, tool_count).
  - Live API tests: end-to-end against Anthropic (guarded by ANTHROPIC_API_KEY, skipped in CI) covering tool use, parallel calls, long chats, large tool results, and cache invalidation.

- **Bug Fixes**
  - Non-mutating transform: clone message content blocks before adding cache_control.
  - Telemetry: use void posthogService.capture to avoid unhandled async.
  - Tests: fix duplicate assistant messages in Scenario 3; add unit test asserting no input mutation; update adapter tests to expect cache_control on user content; add casts for Anthropic cache_read_tokens fields.
  - CI/Vitest: exclude *.live.test.ts and preserve default excludes via configDefaults.exclude.

<sup>Written for commit a754a33527a8808de60b71791032c3ac633a8a31. Summary will update on new commits.</sup>

### CODE REVIEW

File: packages/openai-adapters/src/apis/AnthropicCachingStrategies.test.ts [hunk 1]
```diff
@@ -210,6 +210,123 @@ describe("AnthropicCachingStrategies", () => {
       expect(result.messages).toEqual(body.messages);
     });
 
+    it("should add cache_control to last two user messages for turn-level caching", () => {
```
[the remark this answers is not in the document]
Comment (sestinj) [reply packages/openai-adapters/src/apis/AnthropicCachingStrategies.test.ts:213]: Acknowledged. In practice the total is always exactly 4 breakpoints: 1 system block + 1 last tool + 2 user messages. The system message is always a single-element array (converted from string by `_convertToCleanAnthropicBody`), and we only cache the last tool definition. This matches the VS Code extension path which uses the same `addCacheControlToLastTwoUserMessages` function. Updated the comment in b690c89 to document the breakpoint budget explicitly.

File: extensions/cli/src/stream/streamChatResponse.helpers.ts [hunk 2]
```diff
@@ -377,6 +377,28 @@ export function recordStreamTelemetry(options: {
...
+    const cacheReadTokens =
+      fullUsage.prompt_tokens_details.cache_read_tokens ?? 0;
+    const cacheWriteTokens =
+      fullUsage.prompt_tokens_details.cache_write_tokens ?? 0;
+    const totalPromptTokens = fullUsage.prompt_tokens ?? 0;
+    const cacheHitRate =
+      totalPromptTokens > 0 ? cacheReadTokens / totalPromptTokens : 0;
+
+    try {
+      posthogService.capture("prompt_cache_metrics", {
```
[the remark this answers is not in the document]
Comment (sestinj) [reply extensions/cli/src/stream/streamChatResponse.helpers.ts]: Fixed in b690c89 — added `void` prefix to `posthogService.capture()`.

File: packages/openai-adapters/src/apis/AnthropicCachingStrategies.test.ts [hunk 3]
```diff
@@ -210,6 +210,123 @@ describe("AnthropicCachingStrategies", () => {
...
+          {
+            type: "text",
+            text: "third user message",
+            cache_control: { type: "ephemeral" },
+          },
+        ],
+      });
+    });
+
+    it("should not add cache_control to string content user messages", () => {
```
[the remark this answers is not in the document]
Comment (sestinj) [reply packages/openai-adapters/src/apis/AnthropicCachingStrategies.test.ts:292]: Not an issue in practice. By the time the caching strategy runs, `_convertToCleanAnthropicBody()` has already converted all message content to array format. The string check in `addCacheControlToLastTwoUserMessages` is a defensive guard — string content messages never reach this code path in the real CLI/extension flow. The test documents this edge case behavior explicitly.

File: packages/openai-adapters/src/apis/AnthropicCachingStrategies.test.ts [hunk 4]
```diff
@@ -210,6 +210,123 @@ describe("AnthropicCachingStrategies", () => {
...
+          },
+          {
+            role: "user",
+            content: [{ type: "text", text: "third user message" }],
+          },
+        ],
+        ...body_params,
+      };
+
+      const result = CACHING_STRATEGIES.systemAndTools(body);
```
[the remark this answers is not in the document]
Comment (sestinj) [reply packages/openai-adapters/src/apis/AnthropicCachingStrategies.test.ts:242]: Good catch. Fixed in b690c89 — messages and their content blocks are now cloned before calling `addCacheControlToLastTwoUserMessages`. Added a unit test (`should not mutate the original input body`) that verifies the original content blocks remain unmodified.

File: packages/openai-adapters/src/test/anthropic-caching-scenarios.live.test.ts [hunk 5]
```diff
@@ -0,0 +1,1396 @@
...
+              total: totalInputTokens(usage),
+            });
+
+            // After the first user turn creates cache, subsequent turns should read
+            if (i > 1) {
+              expect(details?.cache_read_tokens).toBeGreaterThan(0);
+              lastRate = rate;
+            }
+
+            // Use actual response as next assistant message
```
[the remark this answers is not in the document]
Comment (sestinj) [reply packages/openai-adapters/src/test/anthropic-caching-scenarios.live.test.ts]: Good catch. Fixed in b690c89 — restructured the loop so that only the first pre-written assistant exchange (i=0) is pushed. For subsequent assistant turns, the real API response from the previous iteration is used instead, avoiding duplicate consecutive assistant messages.

File: packages/openai-adapters/src/test/anthropic-caching-scenarios.live.test.ts [hunk 6]
```diff
@@ -0,0 +1,1405 @@
...
+          // After the first user turn creates cache, subsequent turns should read
+          if (i > 1) {
+            expect(details?.cache_read_tokens).toBeGreaterThan(0);
+            lastRate = rate;
+          }
+
+          // Use actual response as next assistant message
+          messages.push({
+            role: "assistant",
+            content: response.choices[0].message.content ?? "Done.",
```
[the remark this answers is not in the document]
Comment (sestinj) [reply packages/openai-adapters/src/test/anthropic-caching-scenarios.live.test.ts:956]: The `exchanges[i + 1]` fallback was intentionally removed because after the loop refactor, we no longer index into `exchanges` for assistant responses — we always use the real API response. The fallback to `"Done."` is just a safety net for the unlikely case of a null API response.

File: packages/openai-adapters/vitest.config.ts [hunk 7]
```diff
@@ -6,6 +6,7 @@ export default defineConfig({
     globals: true,
     testTimeout: 10000,
     include: ["src/**/*.test.{ts,tsx}", "src/**/*.vitest.{ts,tsx}"],
+    exclude: ["src/**/*.live.test.{ts,tsx}", "node_modules"],
```
[the remark this answers is not in the document]
Comment (sestinj) [reply packages/openai-adapters/vitest.config.ts]: Good catch. Fixed in a754a33 — now spreads `configDefaults.exclude` from `vitest/config` to preserve all default ignores (`dist`, `node_modules`, etc.).
